spring-messaging/src/main/java/org/springframework/messaging/simp/broker/DefaultSubscriptionRegistry.java

```diff
@@ -142,14 +146,17 @@ public String getSelectorHeaderName() {
 		return this.selectorHeaderName;
 	}
 
-
 	@Override
-	protected void addSubscriptionInternal(
-			String sessionId, String subsId, String destination, Message<?> message) {
-
+	protected void addSubscriptionInternal(@NonNull String sessionId, @NonNull String subscriptionId,
+											@NonNull String destination, @NonNull Message<?> message) {
 		Expression expression = getSelectorExpression(message.getHeaders());
-		this.subscriptionRegistry.addSubscription(sessionId, subsId, destination, expression);
-		this.destinationCache.updateAfterNewSubscription(destination, sessionId, subsId);
+		boolean isAntPattern = this.pathMatcher.isPattern(destination);
+		Subscription subscription = new Subscription(subscriptionId, expression, destination, isAntPattern);
+
+		Subscription previousValue = this.subscriptionRegistry.addSubscription(sessionId, subscriptionId, subscription);
+		if (previousValue == null) {
+			this.destinationCache.updateAfterNewSubscription(destination, isAntPattern, sessionId, subscriptionId);
+		}
```

Comment:
Subscription id's are supposed to be unique within a session/connection. That means previousValue should always be null. Do you know of a case where it wouldn't be?

Reply:
I had "better to be safe than sorry" on my mind. Actually, you are right and it complicated the code. I removed it. Thanks for this finding.